core/admin/mailu/api/common.py

```diff
@@ -0,0 +1,33 @@
+from .. import models, utils
+from . import v1
+from flask import request
+import flask
+import hmac
+from functools import wraps
+from flask_restx import abort
+
+def fqdn_in_use(*names):
+    for name in names:
+        for model in models.Domain, models.Alternative, models.Relay:
+            if model.query.get(name):
+                return model
```

Comment:
This is the least efficient implementation I can think of... and I'm confused about the purpose: why don't we return a boolean?

I suggest we either build a proper query and join all the tables... or alternatively map() the relevant get()s and return on first hit.

Reply:
It is a helper function, that is used to check if a domain is already used.
When creating a new item with a new domain, we must check if there is not already a Domain, Alternative or Relayed domain with the same name. This helper function helps checking this. It is used in 5 locations. For example:
https://github.com/Diman0/Mailu/blob/0875975e32a661c246b215e6232e141eda3f0719/core/admin/mailu/api/v1/domains.py#L100

Reply:
I have found it useful to print() the request you ask sqlalchemy to make. A Naked get() like this is a 'SELECT * FROM' whereas ideally you want a 'WHERE EXISTS' query with all the tables and all the names in a single query.

Reply:
Okay, I've googled a bit and this is how it's done:
```
from sqlalchemy.sql.expression import label

def fqdn_in_use(*names):
    d = models.db.session.query(label('name', models.Domain.name))
    a = models.db.session.query(label('name', models.Alternative.name))
    r = models.db.session.query(label('name', models.Relay.name))
    u = d.union_all(a).union_all(r)
    for name in names:
        if u.filter_by(name=name).count():
            return True
    return False
```

@Diman0 could you include this?

We can probably use this at many other locations in mailu.

Reply:
.exists() instead of count(), no?

Reply:
You may even be able to get rid of the for loop and make a single query: with .filter(or_(*search_args)) ... and you'd build search_args potentially without the union

Reply:
Yes, wanted to write this. 
I'll improve it. Just one query is way better.

Reply:
In the api blueprint, only a single name must be checked. So instead of *names, the function could require name as parameter and we could rewrite is as:
```
from sqlalchemy.sql.expression import label

def fqdn_in_use(name):
    d = models.db.session.query(label('name', models.Domain.name))
    a = models.db.session.query(label('name', models.Alternative.name))
    r = models.db.session.query(label('name', models.Relay.name))
    if d.union_all(a).union_all(r).filter_by(name=name).count():
        return True
    return False
```

Reply:
Well, if we don't need to support checking for multiple names we can do simpler and three queries isn't that bad :)

Please don't use .count() as that makes it iterate over all the data/indexes; use .exists() which will bail out on first match found.

Reply:
this uses exists:
```
from sqlalchemy.sql.expression import label

def fqdn_in_use(name):
    d = models.db.session.query(label('name', models.Domain.name))
    a = models.db.session.query(label('name', models.Alternative.name))
    r = models.db.session.query(label('name', models.Relay.name))
    u = d.union_all(a).union_all(r).filter_by(name=name)
    if models.db.session.query(u.exists()).scalar():
        return True
    return False
```

Reply:
Thank you. I will use this.